src/sentry/models/apitoken.py

```diff
@@ -83,12 +84,18 @@ def handle_async_replication(self, region_name: str, shard_identifier: int) -> N
         )
 
     @classmethod
-    def from_grant(cls, grant):
+    def from_grant(cls, grant: ApiGrant):
         with transaction.atomic(router.db_for_write(cls)):
-            return cls.objects.create(
+            api_token = cls.objects.create(
                 application=grant.application, user=grant.user, scope_list=grant.get_scopes()
             )
 
+            # remove the ApiGrant from the database to prevent reuse of the same
+            # authorization code
+            grant.delete()
```

Comment:
do we want to delete this forever, or do we want to soft delete it so we have some sort of trail/audit for at least a few days to help debug and support customers?

Reply:
there's a pipeline that deletes objects/rows after 30 days that we can just use to help clean up

Reply:
The `ApiGrant` needs to be unusable as soon as it is exchanged for a token. We can soft delete these, but I'd like to merge this as it resolves an immediate security bug. I can issue new PRs for the necessary model changes to support soft deletions.

I'll take a look at the cleanup script. It should be removing these that are expired regardless of how old they actually are. Ironically, the cleanup script purging these might have been created to deal with the fallout of this bug. 🤔

Reply:
Yeah, sure thing

Reply:
Ah it does delete them based on the expiring time. 😄 

https://github.com/getsentry/sentry/blob/485f3a42f9a977d4e23b93b4580ad2837ba1027f/src/sentry/runner/commands/cleanup.py#L251-L268